### PR TITLE
Implement Load group

### PR DESCRIPTION
**Pull request by TJStienstra**
Implement #47 

Tasks:
- [x] Fix the problem that the define method of load groups are currently called before the parent has done his own definition
- [x] Write tests

### CODE REVIEW

**Comment by TJStienstra** [comment]
Wanting to call the define methods afterwards is a painful problem I've already messed around with before. I was in fact happy I did not need it anymore. The main problem is around inheritance and having the define methods not getting called multiple times. This can of course be blocked by only letting these define functions run once (forced). However the problem remains that the define method of the load group will still be called in the super.

**Comment by TJStienstra** [comment]
Here is a commit implementing some of that nasty stuff: 02870399a4a5e3155bb1333323d3c3d270b2d6c7

File: src/brim/core/model_base.py [hunk 1]
```diff
@@ -195,6 +195,49 @@ def get_param_values(self, bicycle_parameters: Bicycle) -> dict[Symbol, float]:
...
+    @property
+    def parent(self) -> ModelBase | ConnectionBase | None:
+        """Parent model or connection."""
+        return self._parent
+
+    @parent.setter
+    def parent(self, parent: ModelBase | ConnectionBase) -> None:
+        if self._parent is not None:
+            raise ValueError(f"Load group is already used by {self.parent}")
+        elif not isinstance(parent, self.__annotations__["parent"]):
```
Comment (TJStienstra) [inline src/brim/core/model_base.py:236]: This won't work, optionally I can use `typeguard.check_type`

Comment (TJStienstra) [reply src/brim/core/model_base.py]: Done in 8cf24cffc42055267c844bfe062cd49b9bbf3d4d

**Comment by TJStienstra** [comment]
Just spoke with @brocksam about the problem of the traversal. An option would be to only let `ModelBase` implement `define_objects`:
```py
for submodel in self.submodels:
    submodel.define_objects()
self._define_objects()
for loadgroup in self.load_groups:
    load_group.define_objects()
```
And let every model which inherits `ModelBase` implement `_define_objects` or `define_own_objects`.

**Comment by TJStienstra** [comment]
Decided to rename all define methods to be implemented by subclasses to `_define_...` for clarity sakes, though it would only have been necessary for the models and also connections.

**Comment by brocksam** [comment]
So the user implements the `_define_...` methods when they subclass?

**Comment by TJStienstra** [comment]
> So the user implements the `_define_...` methods when they subclass?

Yes

**Comment by TJStienstra** [comment]
Found `type_guard.check_type` not reliable enough so will be changing it to a similar requirement trick as in models and connections. Though it can this time simply be done in the setter.

**Pull request merged**